### PR TITLE
ENH first pass at camb

### PR DESCRIPTION
**Pull request by beckermr**
This PR has a first pass at CAMB.

### CODE REVIEW

**Comment by beckermr** [comment]
Okay @cmbant. Working builds of pycamb and camb are here! Sorry for the delay. I had to figure out some weird gfortran compatibility issues. The full test suite runs when I build locally as well. I included a shorter test in the package.

This PR is fairly old now so I am pinging the staged recipes maintainers to merge it! 

They are rather busy so it will probably be a bit before this happens.

@conda-forge/staged-recipes

**Comment by cmbant** [comment]
Thanks, looks good.

What was the issue with -ffast-math? 

If the c_loc() change was to enable gfortran 4.8 support, I can merge that change into master. I'd also be happy to merge the inidriver.f90 and modules.f90 changes if we put the conda-specific code in a test for a conda build environment variable (so normally it is ignored).

Btw, intel binary builds run about 30% faster than gfortran, but I don't suppose it's easy to automated that.

Note the licence is not just LGPL, but a modified LGPL.

Is there a way to automate updating this as part of CAMB's travis deploy process (or will it trigger automatically from tagged commits?).

It would also be a good idea to test this with devel branch, since the intention is to move that to master fairly soon (then requires gfortran 6+).

**Comment by beckermr** [comment]
>What was the issue with -ffast-math?

Two things:

1. I was told by a dev that it did not play nice with numpy, but looking at it now, I don't think we are linking with numpy anyhow, so I think we can add it back.
2. The required .o file is not in the libgfortran sources by default so the build was failing locally. Looking at the CI though it appears to be picking up gcc which has the file. I will add the flags back. 

>If the c_loc() change was to enable gfortran 4.8 support, I can merge that change into master. 

Sounds good! I was planning to make a PR but had not gotten to it. I'm going to keep the patch here for now and then when we bump versions we can remove it. 

> I'd also be happy to merge the inidriver.f90 and modules.f90 changes if we put the conda-specific code in a test for a conda build environment variable (so normally it is ignored).

Sure! I will add those in as well.

>Btw, intel binary builds run about 30% faster than gfortran, but I don't suppose it's easy to automated that.

Yeah. Automated distribution for ifort is not going to happen any time soon. :(

>Note the licence is not just LGPL, but a modified LGPL.

Sorry! I will change it. Good catch!

>Is there a way to automate updating this as part of CAMB's travis deploy process (or will it trigger automatically from tagged commits?).

So the way this works is that once the recipe get's merged, a feedstock repo is built. Then conda-forge has a bot that checks the versions of the packages and makes PRs to update them. So once you push an update to pypi, the bot should catch the change and make a PR on the repo. This may not work all the time, so we may need to do things manually. I suppose you could add your own PR automatically when you push to pypi as well. You will have write permissions on the feedstock, so either way it will be easy to update. Generally though, each build needs a human to look at it because things fail in weird ways.

>It would also be a good idea to test this with devel branch, since the intention is to move that to master fairly soon (then requires gfortran 6+).

So this will def fail. The highest gfortran version is 4.8.5. I was working on an update to this and the conda-forge devs know it needs to happen. Building gfortran for conda is a non-trivial task. I was able to get one build to work, but this needs the attention of the core devs who are quite busy right now. Timescales are uncertain, but the next few months is not yet wrong for them to take a look. ;p

**Comment by beckermr** [comment]
Here is an example feedstock repo BTW. 

https://github.com/conda-forge/galsim-feedstock

So version updates would go to the camb one, not here.

**Comment by beckermr** [comment]
The code doesn't link against python AFAICT. Instead the fortran `iso_c_binding` module is used and then the fortran routines are accessed via `c_types`. This should be safe, right?

**Comment by isuruf** [comment]
I don't think so. It's still loading the library. A simple check is,

1. start a new python session. run `print(1e-323)`
2. Do some computations with pycamb
3. run `print(1e-323)`

If 1 and 3 give you different answers, then your library is affecting python.

**Comment by beckermr** [comment]
Ahhhh ok. I will investigate. Thanks for the help!

**Comment by beckermr** [comment]
Hmmmmm. Here is what I got.

The test script is:

```python
#!/use/bin/env python
import camb

# test for weirdness
print("start:", 1e-323)

# run the code
pars = camb.CAMBparams()
pars.set_cosmology(
    H0=67.5, ombh2=0.022, omch2=0.122, mnu=0.06, omk=0, tau=0.06)
pars.InitPower.set_params(As=2e-9, ns=0.965, r=0)
pars.set_for_lmax(2500, lens_potential_accuracy=0)
pars.set_matter_power(redshifts=[0.0], kmax=2.0)
results = camb.get_results(pars)
results.calc_power_spectra(pars)

# compute sigma8 and make sure it is ok
sigma8 = results.get_sigma8()
assert sigma8 > 0.80 and sigma8 < 0.805

print("end:", 1e-323)
```

It prints out during the build:

```
+ python test_camb.py
start: 0.0
end: 0.0
```

The fast-math flag was def used during compilation:

```
running build
Compiling source...
mkdir -p Releaselib
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c constants.f90 -o Releaselib/constants.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c utils.F90 -o Releaselib/utils.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c subroutines.f90 -o Releaselib/subroutines.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c inifile.f90 -o Releaselib/inifile.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c power_tilt.f90 -o Releaselib/power_tilt.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c recfast.f90 -o Releaselib/recfast.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c reionization.f90 -o Releaselib/reionization.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c modules.f90 -o Releaselib/modules.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c bessels.f90 -o Releaselib/bessels.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c equations.f90 -o Releaselib/equations.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c halofit_ppf.f90 -o Releaselib/halofit_ppf.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c lensing.f90 -o Releaselib/lensing.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c SeparableBispectrum.F90 -o Releaselib/SeparableBispectrum.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c cmbmain.f90 -o Releaselib/cmbmain.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c camb.f90 -o Releaselib/camb.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ -c camb_python.f90 -o Releaselib/camb_python.o
gfortran -shared -fPIC -O3 -fopenmp -ffast-math -fmax-errors=4 -JReleaselib -IReleaselib/ Releaselib/constants.o  Releaselib/utils.o   Releaselib/subroutines.o Releaselib/inifile.o  Releaselib/power_tilt.o  Releaselib/recfast.o Releaselib/reionization.o Releaselib/modules.o Releaselib/bessels.o Releaselib/equations.o Releaselib/halofit_ppf.o Releaselib/lensing.o Releaselib/SeparableBispectrum.o Releaselib/cmbmain.o Releaselib/camb.o Releaselib/camb_python.o -o Releaselib/camblib.so
```

Thoughts @isuruf?

**Comment by isuruf** [comment]
What do you get for,
```python
#!/use/bin/env python
print("start:", 1e-323)
import camb
print("after-import:", 1e-323)
# run the code
pars = camb.CAMBparams()
pars.set_cosmology(
    H0=67.5, ombh2=0.022, omch2=0.122, mnu=0.06, omk=0, tau=0.06)
pars.InitPower.set_params(As=2e-9, ns=0.965, r=0)
pars.set_for_lmax(2500, lens_potential_accuracy=0)
pars.set_matter_power(redshifts=[0.0], kmax=2.0)
results = camb.get_results(pars)
results.calc_power_spectra(pars)

# compute sigma8 and make sure it is ok
sigma8 = results.get_sigma8()
assert sigma8 > 0.80 and sigma8 < 0.805

print("end:", 1e-323)
```
?

**Comment by beckermr** [comment]
Ahhhhhhhhhh! 🤦‍♂️ 

Well that was dumb on my part! Checking now...

**Comment by beckermr** [comment]
Yuuuuup. We should def remove the `-ffast-math` flag for the python version. Thanks a bunch @isuruf !

```
+ python test_camb.py
start: 1e-323
after-import: 0.0
end: 0.0
```

**Comment by cmbant** [comment]
Thanks for the updates, sounds good. (Not obvious to me the appveyor build has ever done anything, is there a way to test the Windows conda install at this stage?)

**Comment by beckermr** [comment]
Ahhhh. The windows version is not done yet. My plan is to merge these and then add the windows version afterwards. That build is going to be finicky and I wanted to get the linux and osx ones unblocked first.

File: .circleci/build_steps.sh [hunk 1]
```diff
@@ -40,7 +40,7 @@ conda clean --lock
 # Ensure that `noarch` exists otherwise `conda` won't think this channel is valid.
 conda index /home/conda/staged-recipes/build_artifacts/noarch
 
-conda install --yes --quiet conda-forge-ci-setup=1.* conda-forge-pinning networkx conda-build
+conda install --yes --quiet conda-forge-ci-setup=1.* conda-forge-pinning networkx conda-build>=3.16
```
Comment (isuruf) [inline .circleci/build_steps.sh]: Why is this needed?

Comment (beckermr) [reply .circleci/build_steps.sh]: It isn’t. Leftover from an old version. Conda smithy will overwrite it anyways when the recipe is merged.

**Comment by beckermr** [comment]
Looks like appveyor finally ran. This recipe is ready to be merged!

File: .circleci/build_steps.sh (hunk 1 above)
Comment (beckermr) [reply .circleci/build_steps.sh]: Ack. Now I realize that this is probably needed. Conda build 3.15.1 has a bug where it deletes previous versions of packages when it builds matrices of versions but doesn’t rebuild them later. @msarahan fixed this in 3.16+.

Comment (beckermr) [reply .circleci/build_steps.sh]: The deleted packages are the ones not versioned by the build matrix. That would be the executable in this case.

Comment (beckermr) [reply .circleci/build_steps.sh]: We had to put a similar hack into xgboost when updating that feedstock.

**Comment by beckermr** [comment]
@isuruf @conda-forge/staged-recipes Can we merge this one?

**Comment by beckermr** [comment]
bump on this PR again @isuruf @conda-forge/staged-recipes

File: recipes/camb/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      files:
+        - params.ini
+      commands:
+        - camb params.ini
+
+about:
+  home: https://camb.info/
+  license: modified LGPL
+  license_family: LGPL
+  license_file: {{ RECIPE_DIR }}/LICENSE
```
[the remark this answers is not in the document]
Comment (beckermr) [reply recipes/camb/meta.yaml]: The source at this version did not have a license file. @cmbant has added a license for the source to the repo and so I am including it by hand. Once the version gets bumped (which is blocked by gfortran 6+ being in conda), we can pull it from the source.

**Comment by beckermr** [comment]
OK @isuruf. We good to go?

**Comment by beckermr** [comment]
Thank you!!!!!1 Huzaahhhhhh!!!!!
